interfaces/builtin/userns_test.go

```diff
@@ -0,0 +1,120 @@
+	"github.com/snapcore/snapd/interfaces/builtin"
+	"github.com/snapcore/snapd/interfaces/seccomp"
+	apparmor_sandbox "github.com/snapcore/snapd/sandbox/apparmor"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/testutil"
+)
+
+const plugSnapInfoYaml = `name: consumer
+version: 0
+
+apps:
+  app:
+    command: foo
+    plugs: [userns]
+`
+
+const userNSCoreYaml = `name: core
+version: 0
+type: os
+slots:
+  userns:
+`
+
+type UserNSInterfaceSuite struct {
+	iface interfaces.Interface
+
+	slotInfo *snap.SlotInfo
+	slot     *interfaces.ConnectedSlot
+	plugInfo *snap.PlugInfo
+	plug     *interfaces.ConnectedPlug
+}
+
+var _ = Suite(&UserNSInterfaceSuite{
+	iface: builtin.MustInterface("userns"),
+})
+
+func (s *UserNSInterfaceSuite) SetUpTest(c *C) {
+	s.plug, s.plugInfo = MockConnectedPlug(c, plugSnapInfoYaml, nil, "userns")
+	s.slot, s.slotInfo = MockConnectedSlot(c, userNSCoreYaml, nil, "userns")
+}
+
+func (s *UserNSInterfaceSuite) TestName(c *C) {
+	c.Assert(s.iface.Name(), Equals, "userns")
+}
+
+func (s *UserNSInterfaceSuite) TestSanitizePlug(c *C) {
+	c.Assert(interfaces.BeforePreparePlug(s.iface, s.plugInfo), IsNil)
+}
+
+func (s *UserNSInterfaceSuite) TestNoAppArmor(c *C) {
+	// Ensure that the interface does not fail if AppArmor is unsupported
+	restore := apparmor_sandbox.MockLevel(apparmor_sandbox.Unsupported)
+	defer restore()
+
+	c.Check(interfaces.BeforePreparePlug(s.iface, s.plugInfo), IsNil)
+}
+
+func (s *UserNSInterfaceSuite) TestFeatureDetection(c *C) {
+	// Ensure that the interface does not fail if the userns feature is not present
+	restore := apparmor_sandbox.MockFeatures(nil, nil, nil, nil)
+	defer restore()
+	c.Check(interfaces.BeforePreparePlug(s.iface, s.plugInfo), IsNil)
+}
+
+func (s *UserNSInterfaceSuite) TestAppArmorSpec(c *C) {
+	spec := &apparmor.Specification{}
+	restore := apparmor_sandbox.MockFeatures(nil, nil, []string{"userns"}, nil)
+	defer restore()
+	c.Assert(spec.AddConnectedPlug(s.iface, s.plug, s.slot), IsNil)
+	c.Assert(spec.SecurityTags(), DeepEquals, []string{"snap.consumer.app"})
+	c.Assert(spec.SnippetForTag("snap.consumer.app"), testutil.Contains, "userns,\n")
+}
+
+func (s *UserNSInterfaceSuite) TestSeccompSpec(c *C) {
+	spec := &seccomp.Specification{}
+	c.Assert(spec.AddConnectedPlug(s.iface, s.plug, s.slot), IsNil)
+	c.Assert(spec.SecurityTags(), DeepEquals, []string{"snap.consumer.app"})
+	c.Assert(spec.SnippetForTag("snap.consumer.app"), testutil.Contains, "unshare\n")
+}
+
+func (s *UserNSInterfaceSuite) TestInterfaces(c *C) {
+	c.Check(builtin.Interfaces(), testutil.DeepContains, s.iface)
+}
+
+func (s *UserNSInterfaceSuite) TestAutoConnect(c *C) {
+	c.Assert(s.iface.AutoConnect(nil, nil), Equals, true)
+}
+
+func (s *UserNSInterfaceSuite) TestStaticInfo(c *C) {
+	si := interfaces.StaticInfoOf(s.iface)
+	c.Check(si.ImplicitOnCore, Equals, false)
+	c.Check(si.ImplicitOnClassic, Equals, false)
+	c.Check(si.Summary, Equals, `allows the ability to use user namespaces`)
```

Comment:
this needs changes

Reply:
also done in https://github.com/snapcore/snapd/commit/7219f5a16653cc745c971149cb044e71c040461a